runtime/memprof.c

```diff
@@ -0,0 +1,281 @@
+    beta_rec = sqrt((3./(PI*PI))*cur_lambda);
+    k = logf(c*beta_rec) - cur_lambda;
+    while(1) {
+      double u, n, v, y;
+      u = mt_generate_uniform();
+      y = log(1./u-1);
+      n = floor(cur_lambda - y*beta_rec + 0.5);
+      if(n < 0.)
+        continue;
+      v = mt_generate_uniform();
+      if(y + log(v*u*u) < k + n*log_cur_lambda - lfact(n))
+        return n > ((1<<30)-2) ? ((1<<30)-2) : n;
+    }
+  }
+}
+
+/**** Interface with the OCaml code. ****/
+
+CAMLprim value caml_memprof_set(value v) {
+  CAMLparam1(v);
+  double l = Double_val(Field(v, 0));
+  intnat sz = Long_val(Field(v, 1));
+
+  if(sz < 0 || !(l >= 0.) || l > 1.)
+    caml_failwith("caml_memprof_set");
+
+  lambda = l;
+
+  callstack_size = sz;
+  memprof_callback = Field(v, 2);
+
+  if(!init) {
+    int i;
+    mt_index = 624;
+    mt_state[0] = 42;
+    for(i = 1; i < 624; i++)
+      mt_state[i] = 0x6c078965 * (mt_state[i-1] ^ (mt_state[i-1] >> 30)) + i;
+
+    caml_register_global_root(&memprof_callback);
+  }
+
+  CAMLreturn(Val_unit);
+}
+
+/* Cf. Gc.Memprof.alloc_kind */
+enum ml_alloc_kind {
+  Minor = Val_long(0),
+  Major = Val_long(1),
+  Serialized = Val_long(2)
+};
+
+static value do_callback(tag_t tag, intnat wosize, int32_t occurences,
+                         value callstack, enum ml_alloc_kind cb_kind) {
+  CAMLparam1(callstack);
+  CAMLlocal1(sample_info);
+  CAMLassert(occurences > 0);
+
+  sample_info = caml_alloc_small(5, 0);
+  Field(sample_info, 0) = Val_long(occurences);
+  Field(sample_info, 1) = cb_kind;
+  Field(sample_info, 2) = Val_long(tag);
+  Field(sample_info, 3) = Val_long(wosize);
+  Field(sample_info, 4) = callstack;
+
+  CAMLreturn(caml_callback_exn(memprof_callback, sample_info));
+}
+
+void caml_memprof_set_suspended(int new_suspended) {
+  caml_memprof_suspended = new_suspended;
+}
+
+/**** Sampling procedures ****/
+
+static value capture_callstack(int avoid_gc) {
+  return caml_get_current_callstack_impl(callstack_size, avoid_gc);
+}
+
+struct caml_memprof_postponed_block {
+  value block;
+  value callstack;
+  int32_t occurences;
+  struct caml_memprof_postponed_block* next;
+} *caml_memprof_postponed_head = NULL;
+
+/* When allocating in the major heap, we cannot call the callback, because
+   [caml_alloc_shr] is guaranteed not to call the GC. Hence, this function
+   determines if the block need to be sampled, and if so, it registers the
+   block in the todo-list so that the callback call is performed when
+   possible. */
+void caml_memprof_track_alloc_shr(value block) {
+  int32_t occurences = mt_generate_poisson(Whsize_val(block));
+  CAMLassert(Is_in_heap(block));
+  if(occurences > 0) {
+    struct caml_memprof_postponed_block* pb =
+      caml_stat_alloc_noexc(sizeof(struct caml_memprof_postponed_block));
+    if(pb == NULL) return;
+    pb->block = block;
+    caml_register_generational_global_root(&pb->block);
+    caml_memprof_set_suspended(1);
+    pb->callstack = capture_callstack(1);
```

Comment:
I don't have a strong argument. but: if you let the user gather a call stack, they will get a callstack for the place in their code where they implemented the monitoring logic, rather than a callstack for the allocation point. The latter is a sub-stack of the former (given that the monitor is called from the allocation point), but that remains a less pleasant user-interface with some irrelevant frames at the top -- for example, "the location at the top of the stack" does not locate the allocation point anymore.

Reply:
> This patch adds an extra parameter to caml_get_current_callstack in order to support this call from within caml_alloc_shr. Why not collect the callstack when the callback runs, so that this extra parameter is not needed?
>
> In fact, why collect the callstack at all in this code? The callback can choose to call Printexc.get_callstack itself. That way, we can also remove the callstack_size global variable and the callstack_size field in Gc.Memprof.ctrl.

If you are annoyed by extra parameter, the we can decide to always allocate callstacks in the major heap. I am not so sure of the performance implications of this, though.

The reason is that if you take the call stack when calling the callback, then the callstack is going to be different, and will not reflect the precise point of allocation, since the call to the callback might have been postponed.

A similar issue reappears in the context of combined allocations in the minor heap: If you let the user take their own callstack when the callback is called, then there is no way of distinguishing which of the allocated block is sampled. This is even worse if the callback is called during minor GC, as you proposed recently.

Now, I understand there is a choice to do here: are we OK with approximate callstacks, or would we prefer to make sure that the first callstack slot represents the point of allocation? I tried to make sure of the latter, with some efforts. And I think this is better, since blocks allocated very closely can have very different behaviors wrt. liveness.

Reply:
> The reason is that if you take the call stack when calling the callback, then the callstack is going to be different, and will not reflect the precise point of allocation, since the call to the callback might have been postponed.

I do not think that this is true. In every case that I can think of, the backtrace is exactly the same whether the callstack is taken in `caml_alloc_shr` or at the next safepoint (generally `caml_check_urgent_gc`). So I don't think the extra complexity of taking backtraces early and keeping them until the callback runs is buying us any extra accuracy.

Can you think of any situation where the backtraces are actually different?

> If you are annoyed by extra parameter [...]

I am in fact annoyed by the extra parameter, and perhaps it's worth explaining why! One of the most tricky parts of working on the OCaml runtime system is keeping track of which functions can and cannot call the GC, to ensure that the GC's invariants are upheld at any point where it can run. Having a function whose GC status depends on the value of a runtime parameter will make this already tricky reasoning worse.

Reply:
@gasche You're right that having the user take the backtrace will introduce an annoying extra frame, so I retract the idea that we shouldn't collect the bactrace in the runtime. I think the right thing to do is have the runtime collect the backtrace, but at the point of invoking the callback rather than during `caml_alloc_shr`

Reply:
> Can you think of any situation where the backtraces are actually different?

I expect many C bindings for various C external libraries do not properly call `caml_check_urgent_gc` after every call to `caml_alloc_shr`. And even if we can now document this as being "good practice", I do not think libraries will be updated any soon.

Also, note that in bytecode mode, if no call to `caml_check_urgent_gc` is issued, the postonned blocks will only be checked at the next `CHECK_SIGNALS` instruction, at next end of `try .. with ...` block or at the next minor GC.

Appart from that, there are more anecdotic uses of `caml_alloc_shr` in `interp.c` that are not paired with a call to `caml_check_urgent_gc`. This corresponds to, e.g., allocating a large closure, or a large record. I can now add the call, though: it should not hurt performances much, since it is amortized by initialization of the block which has size >= 256.

> Having a function whose GC status depends on the value of a runtime parameter will make this already tricky reasoning worse.

I think that having a parameter which is explicitly dedicated to this purpose makes very clear whether one particular call can trigger the GC or not. If you prefer I can use an `enum` to make that clearer at the point of allocation.

Reply:
Alright, I have just pushed a new version, please let me know what you think about it.

In a nutshell, instead of using a parameter indicating which allocation function to use, we split the `caml_get_current_callstack_impl` in two parts. The first part computes the size of the allocation, while the second copies the call stack to the allocated block. Hence, the choice of the allocation function is done by the caller.

Reply:
> I expect many C bindings for various C external libraries do not properly call caml_check_urgent_gc after every call to caml_alloc_shr. And even if we can now document this as being "good practice", I do not think libraries will be updated any soon.

Ah, yes, C bindings that allocate memory with caml_alloc_shr but do not contain a safepoint will cause the backtrace to be wrong. (A C function that uses any allocation function any time after calling caml_alloc_shr will do the right thing. It doesn't matter where in the C function the safepoint exists, since this patch records OCaml backtraces that do not describe C locations).

However, caml_alloc_shr is used extremely rarely outside the compiler distribution, so I'm not sure these cases are worth special attention.

> Also, note that in bytecode mode, if no call to caml_check_urgent_gc is issued, the postonned blocks will only be checked at the next CHECK_SIGNALS instruction, at next end of try .. with ... block or at the next minor GC.
> 
> Appart from that, there are more anecdotic uses of caml_alloc_shr in interp.c that are not paired with a call to caml_check_urgent_gc. This corresponds to, e.g., allocating a large closure, or a large record. I can now add the call, though: it should not hurt performances much, since it is amortized by initialization of the block which has size >= 256.

Good point! As you say, these cases seem easy to fix in the interpreter.

> I think that having a parameter which is explicitly dedicated to this purpose makes very clear whether one particular call can trigger the GC or not. If you prefer I can use an enum to make that clearer at the point of allocation.

No, an `int` is fine. My concern isn't with how the condition is expressed; it's with the idea of a function whose GC status depends on a parameter.

There is already a mechanism for recording backtraces in a way that does not involve GC, used by `caml_stash_backtrace`. Is it possible to reuse this mechanism instead of reusing the one that does involve GC and making it not involve GC sometimes?

Reply:
I wrote the above comment before realising you'd pushed more code! I'll read the new implementation of backtrace capturing now.

Reply:
I've read the new code, this implementation looks good. I'm happy with the split into two functions. Apologies for the out-of-order comments above!